components/nimbus/src/stateful/gecko_prefs.rs

```diff
@@ -0,0 +1,337 @@
+pub struct PrefEnrollmentData {
+    pub pref_value: PrefValue,
+    pub feature_id: String,
+    pub variable: String,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GeckoPrefState {
+    pub gecko_pref: GeckoPref,
+    pub gecko_value: Option<PrefValue>,
+    pub enrollment_value: Option<PrefEnrollmentData>,
+    pub is_user_set: bool,
+}
+
+impl GeckoPrefState {
+    pub fn new(pref: &str, branch: Option<PrefBranch>) -> Self {
+        Self {
+            gecko_pref: GeckoPref {
+                pref: pref.into(),
+                branch: branch.unwrap_or(PrefBranch::Default),
+            },
+            gecko_value: None,
+            enrollment_value: None,
+            is_user_set: false,
+        }
+    }
+
+    pub fn with_gecko_value(mut self, value: PrefValue) -> Self {
+        self.gecko_value = Some(value);
+        self
+    }
+
+    pub fn with_enrollment_value(mut self, pref_enrollment_data: PrefEnrollmentData) -> Self {
+        self.enrollment_value = Some(pref_enrollment_data);
+        self
+    }
+
+    pub fn set_by_user(mut self) -> Self {
+        self.is_user_set = true;
+        self
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, Eq, PartialEq, Copy)]
+pub enum PrefUnenrollReason {
+    Changed,
+    FailedToSet,
+}
+
+pub type MapOfFeatureIdToPropertyNameToGeckoPrefState =
+    HashMap<String, HashMap<String, GeckoPrefState>>;
+
+pub fn create_feature_prop_pref_map(
+    list: Vec<(&str, &str, GeckoPrefState)>,
+) -> MapOfFeatureIdToPropertyNameToGeckoPrefState {
+    list.iter().fold(
+        HashMap::new(),
+        |mut feature_map, (feature_id, prop_name, pref_state)| {
+            feature_map
+                .entry(feature_id.to_string())
+                .or_default()
+                .insert(prop_name.to_string(), pref_state.clone());
+            feature_map
+        },
+    )
+}
+
+pub trait GeckoPrefHandler: Send + Sync {
+    /// Used to obtain the prefs values from Gecko
+    fn get_prefs_with_state(&self) -> MapOfFeatureIdToPropertyNameToGeckoPrefState;
+
+    /// Used to set the state for each pref based on enrollments
+    fn set_gecko_prefs_state(&self, new_prefs_state: Vec<GeckoPrefState>);
+}
+
+#[derive(Default)]
+pub struct GeckoPrefStoreState {
+    pub gecko_prefs_with_state: MapOfFeatureIdToPropertyNameToGeckoPrefState,
+}
+
+impl GeckoPrefStoreState {
+    pub fn update_pref_state(&mut self, new_pref_state: &GeckoPrefState) -> bool {
+        self.gecko_prefs_with_state
+            .iter_mut()
+            .find_map(|(_, props)| {
+                props.iter_mut().find_map(|(_, pref_state)| {
+                    if pref_state.gecko_pref.pref == new_pref_state.gecko_pref.pref {
+                        *pref_state = new_pref_state.clone();
+                        Some(true)
+                    } else {
+                        None
+                    }
+                })
+            })
+            .is_some()
```

Comment:
This is always either Some(true) or None, so this should change slightly


```suggestion
            .find(|(_, props)| {
                props.iter_mut().find_map(|(_, pref_state)| {
                    if pref_state.gecko_pref.pref == new_pref_state.gecko_pref.pref {
                        *pref_state = new_pref_state.clone();
                        true
                    } else {
                        false
                    }
                })
            })
            .is_some()
```

Reply:
`find` doesn't like working with mutable content, but `find_map` works there. I get `cannot assign a new value to immutable borrowed content` errors with `find`